File: src/main/java/com/mparticle/kits/StringUtils.java
package com.mparticle.kits;

import com.apptentive.android.sdk.ApptentiveLog;

import java.util.Map;

import static com.apptentive.android.sdk.ApptentiveLogTag.UTIL;

final class StringUtils {
    public static boolean tryParseSettingFlag(Map<String, String> settings, String key, boolean defaultValue) {
        final String value = settings.get(key);
        if (value != null) {
            final Boolean flag = StringUtils.tryParseBoolean(value);
            if (flag != null) {
                return flag;
            }
            ApptentiveLog.w(UTIL, "Unable to parse boolean flag '%s': %s", key, value);
        }

        return defaultValue;
    }

    public static Number tryParseNumber(String value) {
        Long longValue = tryParseLong(value);
        if (longValue != null) {
            if (longValue >= Integer.MIN_VALUE && longValue <= Integer.MAX_VALUE) {
                return longValue.intValue();
            }
            return longValue;
        }

        return tryParseDouble(value);
    }

    public static Long tryParseLong(String value) {
        try {
            return Long.valueOf(value);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public static Double tryParseDouble(String value) {
        try {
            return Double.valueOf(value);
        } catch (NumberFormatException e) {
            return null;
        }
    }

    public static Boolean tryParseBoolean(String value) {
        try {
            return Boolean.valueOf(value);
        } catch (NumberFormatException e) {
            return null;
        }
    }
}

File: src/main/kotlin/com/mparticle/kits/StringUtils.kt
package com.mparticle.kits

import android.util.Log

internal object StringUtils {
    @JvmStatic
    fun tryParseSettingFlag(
        settings: Map<String, String>,
        key: String?,
        defaultValue: Boolean
    ): Boolean {
        val value = settings[key]
        if (value != null) {
            val flag = value.toBoolean()
            if (flag != null) {
                return flag
            }
            Log.w(
                "MPartile-Util",
                "Unable to parse boolean flag $key: $value")
        }
        return defaultValue
    }

    @JvmStatic
    fun tryParseNumber(value: String): Number? {
        val longValue = tryParseLong(value)
        return if (longValue != null) {
            if (longValue >= Int.MIN_VALUE && longValue <= Int.MAX_VALUE) {
                longValue.toInt()
            } else longValue
        } else tryParseDouble(value)
    }

    fun tryParseLong(value: String): Long? {
        return try {
            value.toLong()
        } catch (e: NumberFormatException) {
            null
        }
    }

    fun tryParseDouble(value: String): Double? {
        return try {
            java.lang.Double.valueOf(value)
        } catch (e: NumberFormatException) {
            null
        }
    }

    fun tryParseBoolean(value: String): Boolean? {
        return try {
            value.toBoolean()
        } catch (e: NumberFormatException) {
            null
        }
    }
}